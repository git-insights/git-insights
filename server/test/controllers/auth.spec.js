const { mockRequest, mockResponse } = require("../utils/interceptor.js");
const {
  getLoginSuccess,
  getLoginFailure,
  getLogout,
  postGithubAuthenticate
} = require("../../src/controllers/auth.js");

describe("Controller 'getLoginSuccess' ", () => {
  test("should respond with success object, when user exists in request.", async () => {
    const req = mockRequest("../fixtures/get_login_success.json");
    const res = mockResponse();
    await getLoginSuccess(req, res);

    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toMatchSnapshot();
  });
  test("should not respond, when user doesn't exist on request.", async () => {
    const req = mockRequest();
    const res = mockResponse();
    await getLoginSuccess(req, res);

    expect(res.json).toHaveBeenCalledTimes(0);
  });
});

describe("Controller 'getLoginFailure' ", () => {
  test("should respond with 401 status and failure object.", async () => {
    const req = mockRequest();
    const res = mockResponse();

    await getLoginFailure(req, res);

    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(401);
    expect(res.json).toHaveBeenCalledTimes(1);
    expect(res.json).toMatchSnapshot();
  });
});

describe("Controller 'getLogout' ", () => {
  test("should call logout and redirect to client home page.", async () => {
    const req = mockRequest();
    const res = mockResponse();
    req.logout = jest.fn().mockReturnValue(req);
    res.redirect = jest.fn().mockReturnValue(req);

    await getLogout(req, res);

    expect(req.logout).toHaveBeenCalledTimes(1);
    expect(res.redirect).toHaveBeenCalledTimes(1);

    // NOTE: this will break when constant `CLIENT_HOME_PAGE_URL` is used;
    expect(res.redirect).toMatchSnapshot();
  });
});

describe("Controller 'postGithubAuthenticate' ", () => {
  test("should respond with 401 status and failure message, when user doesn't exist in request", async () => {
    const req = mockRequest();
    const res = mockResponse();
    const next = jest.fn();

    await postGithubAuthenticate(req, res, next);

    expect(res.send).toHaveBeenCalledTimes(1);
    expect(res.send).toMatchSnapshot();
    expect(next).toHaveBeenCalledTimes(0);
  });

  test("should set req.auth object with user id and call next, when user exists in request", async () => {
    const req = mockRequest("../fixtures/post_github_authenticate.json");
    const res = mockResponse();
    const next = jest.fn();

    await postGithubAuthenticate(req, res, next);

    expect(res.send).toHaveBeenCalledTimes(0);
    expect(req.auth).toMatchSnapshot();
    expect(next).toHaveBeenCalledTimes(1);
  });
});
