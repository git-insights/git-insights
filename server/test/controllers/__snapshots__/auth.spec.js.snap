// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller 'getLoginFailure'  should respond with 401 status and failure object. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "user failed to authenticate.",
        "success": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "message": "user failed to authenticate.",
                "success": false,
              },
            ],
          ],
          "results": [Circular],
        },
        "send": [MockFunction],
        "status": [MockFunction] {
          "calls": Array [
            Array [
              401,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`Controller 'getLoginSuccess'  should respond with success object, when user exists in request. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "cookies": Object {},
        "message": "user has successfully authenticated",
        "success": true,
        "user": Object {},
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "cookies": Object {},
                "message": "user has successfully authenticated",
                "success": true,
                "user": Object {},
              },
            ],
          ],
          "results": [Circular],
        },
        "send": [MockFunction],
        "status": [MockFunction],
      },
    },
  ],
}
`;

exports[`Controller 'getLogout'  should call logout and redirect to client home page. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://www.google.com",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "body": [MockFunction],
        "logout": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "params": [MockFunction],
      },
    },
  ],
}
`;

exports[`Controller 'postGithubAuthenticate'  should respond with 401 status and failure message, when user doesn't exist in request 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      401,
      "User Not Authenticated",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": [MockFunction],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              401,
              "User Not Authenticated",
            ],
          ],
          "results": [Circular],
        },
        "status": [MockFunction],
      },
    },
  ],
}
`;

exports[`Controller 'postGithubAuthenticate'  should set req.auth object with user id and call next, when user exists in request 1`] = `
Object {
  "id": "USER_ID_TEST",
}
`;
